pkg/api/server/v1alpha2/lister/aggregator.go

```diff
@@ -0,0 +1,200 @@
+	var err error
+	summary := make([]map[string]interface{}, 0)
+	db = db.Model(&tdb.Record{})
+	db, err = a.buildQuery(ctx, db)
+	if err != nil {
+		return nil, err
+	}
+
+	db = a.applyAggregateFunc(ctx, db)
+	db.Scan(&summary)
+
+	sm, err := ToSummary(summary)
+	if err != nil {
+		return nil, err
+	}
+	return sm, nil
+}
+
+func ToSummary(summary []map[string]interface{}) (*pb.Summary, error) {
+	var data []*structpb.Struct
+	for _, s := range summary {
+		m := make(map[string]*structpb.Value)
+		for sk, sv := range s {
+			pbValue, err := structpb.NewValue(sv)
+			if err != nil {
+				return nil, err
+			}
+			m[sk] = pbValue
+		}
+		data = append(data, &structpb.Struct{Fields: m})
+	}
+
+	return &pb.Summary{
+		Summary: data,
+	}, nil
+}
+
+var validGroups = map[string]struct{}{
+	"year":   {},
+	"month":  {},
+	"week":   {},
+	"day":    {},
+	"hour":   {},
+	"minute": {},
+}
+
+func newAggregator(env *cel.Env, aggregateObjectRequest summaryRequest, clauses ...equalityClause) (*Aggregator, error) {
+	filter := &filter{
+		env:             env,
+		expr:            strings.TrimSpace(aggregateObjectRequest.GetFilter()),
+		equalityClauses: clauses,
+	}
+
+	summary := strings.Split(strings.TrimSpace(aggregateObjectRequest.GetSummary()), ",")
+
+	aggregators, err := getAggregateFunc(summary)
+	if err != nil {
+		return nil, err
+	}
+
+	group := strings.TrimSpace(aggregateObjectRequest.GetGroupBy())
+	if group != "" {
+		_, ok := validGroups[group]
+		if !ok {
+			return nil, status.Errorf(codes.InvalidArgument, "group_by does not recognize %s", group)
+		}
+		aggregators = append(aggregators, GroupBy(group, groupQuery))
+	}
+
+	return &Aggregator{
+		aggregators: aggregators,
+		queryBuilders: []queryBuilder{
+			filter,
+		},
+	}, nil
+}
+
+func GetDuration(fn, query, value string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Select(db.Statement.Selects, fn+"("+query+")::INTERVAL as "+value)
+	}
+}
+
+func GroupBy(group, query string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		groupSelect := "EXTRACT(EPOCH FROM DATE_TRUNC('" + group + "', " + query + ")) AS group_value"
+		return db.Select(db.Statement.Selects, groupSelect).Group("group_value")
+	}
+}
+
+func GetStatus(query, reason string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		statusSelect := ""
+		if reason == "Unknown" {
+			statusSelect = "COUNT(CASE WHEN " + query + " NOT IN ('Failed', 'Succeeded', 'Cancelled') THEN 1 END) AS unknown"
+		} else {
+			statusSelect = "COUNT(CASE WHEN " + query + " = '" + reason + "' THEN 1 END) AS " + strings.ToLower(reason)
+		}
+		return db.Select(db.Statement.Selects, statusSelect)
+	}
+}
+
+func GetCount(query, countName string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Select(db.Statement.Selects, "COUNT("+query+") AS "+countName)
+	}
+}
+
+type aggregateFunc func(db *gorm.DB) *gorm.DB
+
+var summaryFuncs = map[string]aggregateFunc{
+	"total":          GetCount("*", "total"),
```

Comment:
Can we just make a single call? Like I did? Why don't we use TYPE as a group also?

Reply:
Then we can get all these in the single call.

Reply:
I actually took the inspiration from your code and then came up with this 😅. Practically, it is a single call only. You can enable DB logs and see them. 

This design lets us define and add new queries without adding a lot of code, but still maintains the code readability. Furthermore, we can use the same function for both records and results and even a combination of many functions.